Architecture.md

```diff
@@ -71,6 +71,116 @@
+
+The git server is written in golang, for its simplicity and its performance. It is also well documented, and has a lot of libraries to help us build the server, unlike Rust which was the original choice. The libraries used are [ssh made by gliderlabs](https://github.com/gliderlabs/ssh) which is used to create git clients and servers. It is also used by [gitea](https://about.gitea.com) to create their git server.
+
+The authentication and authorization process is the following:
+
+- The client connects to the git server using ssh via `git push`
+- The git server checks if the client is authenticated using the ssh key provided by the client by matching it against the ssh keys stored in the database
+    - If the client is not authenticated, the connection is closed
+- The git server checks if the client is authorized to push to the repository by checking if the client is the owner of the repository
+    - If the client is not authorized, the connection is closed
+- The git server executes the command `git-receive-pack` which is used to push to a repository
+
+```mermaid
+sequenceDiagram
+    participant Client
+    participant GitServer
+    participant Database
+    participant HostMachine
+
+    Client->>GitServer: git-send-pack repository_name (git push)
+    GitServer->>Database: Is the ssh key in the database?
+    Database-->>GitServer: result
+    alt Authenticated
+        GitServer->>Database: Is the ssh key linked to the repository ?
+        Database-->>GitServer: result
+        alt Authorized
+            GitServer->>HostMachine: Execute git-receive-pack
+            HostMachine-->>GitServer: Push Successful (from host)
+            GitServer-->>Client: Push Successful (continue)
+        else Not Authorized
+            GitServer-->>Client: Connection Closed (Not Authorized)
+        end
+    else Not Authenticated
+        GitServer-->>Client: Connection Closed (Not Authenticated)
+    end
+```
+
+##### Building images
+
+Once the code is finally pushed to the repository, it needs to be built in order to be deployed. Since the strategy we chose is [Docker](https://www.docker.com/), the code needs to be built into a docker image.
+
+In that essence, since the client code can be written in any language, we need to be able to build the code in any language. To do so, we use [buildpacks](https://buildpacks.io/), which is a tool to build code in any language into a docker image. It is used by [heroku](https://www.heroku.com/) to build their applications.
+
+How can we trigger the build of the image ? For a first version, we chose to use [git hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) to trigger a bash script which will build the image on the host machine. This is not scalable, as the build is done on the host machine. It also poses questions about security risks since our ssh server needs to be able to access the docker daemon on the host machine. But it is a good first step to test the concept.
```

Comment:
```suggestion
How can we trigger the build of the image? For a first version, we chose to use [git hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) to trigger a bash script which will build the image on the host machine.
This is not scalable, as the build is done on the host machine.
It also poses questions about security risks since our ssh server needs to be able to access the docker daemon on the host machine. But it is a good first step to test the concept.
```